azure-kusto-ingest/azure/kusto/ingest/managed_streaming_ingest_client.py

```diff
@@ -0,0 +1,89 @@
+import uuid
+from io import SEEK_SET
+from typing import TYPE_CHECKING, Union, Optional, IO, AnyStr
+
+from azure.kusto.data import KustoConnectionStringBuilder
+from azure.kusto.data.exceptions import KustoApiError
+from . import IngestionProperties, BlobDescriptor, StreamDescriptor, FileDescriptor
+from ._retry import ExponentialRetry
+from ._stream_extensions import read_until_size_or_end, chain_streams
+from .base_ingest_client import BaseIngestClient, IngestionResult
+from .ingest_client import QueuedIngestClient
+from .streaming_ingest_client import KustoStreamingIngestClient
+
+if TYPE_CHECKING:
+    pass
+
+
+class ManagedStreamingIngestClient(BaseIngestClient):
+    MAX_STREAMING_SIZE_IN_BYTES = 4 * 1024 * 1024
+    ATTEMPT_COUNT = 4
+
+    def __init__(self, queued_kcsb: Union[KustoConnectionStringBuilder, str], streaming_kcsb: Optional[Union[KustoConnectionStringBuilder, str]] = None):
+
+        if streaming_kcsb is None:
+            kcsb = repr(queued_kcsb) if type(queued_kcsb) == KustoConnectionStringBuilder else queued_kcsb
+            streaming_kcsb = KustoConnectionStringBuilder(kcsb.replace("https://ingest-", "https://"))
+
+        self.queued_client = QueuedIngestClient(queued_kcsb)
+        self.streaming_client = KustoStreamingIngestClient(streaming_kcsb)
+
+    def ingest_from_file(self, file_descriptor: Union[FileDescriptor, str], ingestion_properties: IngestionProperties) -> IngestionResult:
+        stream_descriptor = StreamDescriptor.from_file_descriptor(file_descriptor)
+
+        with stream_descriptor.stream:
+            return self.ingest_from_stream(stream_descriptor, ingestion_properties)
+
+    def ingest_from_stream(self, stream_descriptor: Union[StreamDescriptor, IO[AnyStr]], ingestion_properties: IngestionProperties) -> IngestionResult:
+        set_client_request_id = False
+        random_uuid = uuid.uuid4()
+        if not ingestion_properties.client_request_id:
+            set_client_request_id = True
+            ingestion_properties.client_request_id = ManagedStreamingIngestClient._get_request_id(random_uuid, 0)
+
+        stream_descriptor = BaseIngestClient._prepare_stream(stream_descriptor, ingestion_properties)
+        stream = stream_descriptor.stream
+
+        buffered_stream = read_until_size_or_end(stream, self.MAX_STREAMING_SIZE_IN_BYTES + 1)
```

Comment:
Won't it be easier just to try the 1st time as is and if the error cods is "Payload too large" (or rest payload exceeds the max size) fallback to queued instead of splitting and chaining?

Reply:
The problem is that the stream might not be rewindable

Reply:
The common use case is to send an uncompressed stream which you will compress anyway - I don't think it worth the trouble to support non seek-able streams unless you want to change it for other SDKs too.

Reply:
In node we have to do it since we don't have seek (There's a PR - https://github.com/Azure/azure-kusto-node/pull/155)
I think the plan is to add it to the other SDKs too

Reply:
@ArielYehezkely , we discussed this problem in general and decided in favor of the above logic for languages without built in seek-able streams